src/main/java/org/folio/linked/data/service/resource/ResourceMarcServiceImpl.java

```diff
@@ -87,26 +92,34 @@ public ResourceMarcViewDto getResourceMarcView(Long id) {
 
   @Override
   public Boolean isSupportedByInventoryId(String inventoryId) {
-    var response = srsClient.getFormattedSourceStorageInstanceRecordById(inventoryId);
-    return Optional.ofNullable(response.getBody())
+    return getRecord(inventoryId)
+      .map(HttpEntity::getBody)
       .map(Record::getParsedRecord)
       .map(parsedRecord -> (Map<?, ?>) parsedRecord.getContent())
       .map(content -> (String) content.get("leader"))
       .map(this::isMonograph)
-      .orElse(false);
+      .orElseThrow(() -> createNotFoundException(inventoryId));
   }
 
   @Override
   public ResourceResponseDto getResourcePreviewByInventoryId(String inventoryId) {
-    var response = srsClient.getFormattedSourceStorageInstanceRecordById(inventoryId);
-    return Optional.ofNullable(response.getBody())
-      .map(Record::getParsedRecord)
-      .map(ParsedRecord::getContent)
-      .map(this::toJsonString)
-      .flatMap(marcBib2ldMapper::fromMarcJson)
+    return getResource(inventoryId)
       .map(resourceModelMapper::toEntity)
       .map(resourceDtoMapper::toDto)
-      .orElse(null);
+      .orElseThrow(() -> createNotFoundException(inventoryId));
+  }
+
+  @Override
+  public ResourceIdDto importMarcRecord(String inventoryId) {
+    return getResource(inventoryId)
+      .map(resource -> {
+        resource.getFolioMetadata().setSource(LINKED_DATA);
+        return resource;
+      })
+      .map(this::saveMarcResource)
```

Comment:
Hi @askhat-abishev,

The current implementation of `saveMarcResource` update/replace the existing resource if resource already exists in the system. However, for this use case, we need to throw an error when the resource already exists in the system. Please see the comment in https://folio-org.atlassian.net/browse/UILD-408

So, we need to call a different version of `saveMarcResource` here with the following logic

```
    if (folioMetadataRepository.existsBySrsId(modelResource.getFolioMetadata().getSrsId())) {
      throw new AlreadyExistsException("MARC record having srsID {} is already imported to data graph");
    }
    
    if (resourceRepo.existsById(modelResource.getId()) || ) {
      throw new AlreadyExistsException("Another resource with ID {} already exists in the graph");
    }
    
    return createResource(mapped);
```

Reply:
done